File: api/tts.py
"""
Vercel Python Serverless Function: POST /api/tts
Body (JSON): { text, voice, rate="+0%", pitch="+0Hz" }
Response: audio/mpeg (binary MP3) - base64 encoded for binary transport
"""
import json
import asyncio
import base64
import traceback

try:
    import edge_tts
except ImportError as e:
    def handler(event, context):
        return {
            "statusCode": 500,
            "headers": {
                "Content-Type": "application/json",
                "Access-Control-Allow-Origin": "*",
                "Access-Control-Allow-Headers": "Content-Type",
                "Access-Control-Allow-Methods": "POST,OPTIONS",
            },
            "body": json.dumps({"error": f"edge-tts not installed ({e}). Add requirements.txt with edge-tts"}),
        }


async def _synthesize(text, voice, rate, pitch):
    communicate = edge_tts.Communicate(text, voice, rate=rate, pitch=pitch)
    audio_bytes = bytearray()
    async for chunk in communicate.stream():
        if chunk.get("type") == "audio":
            audio_bytes.extend(chunk["data"])
    return bytes(audio_bytes)


def handler(event, context):
    method = event.get("httpMethod") or event.get("method", "GET")
    headers = {
        "Access-Control-Allow-Origin": "*",
        "Access-Control-Allow-Headers": "Content-Type",
        "Access-Control-Allow-Methods": "POST,OPTIONS",
    }
    if method == "OPTIONS":
        return {"statusCode": 200, "headers": headers, "body": ""}
    if method != "POST":
        headers["Content-Type"] = "application/json"
        return {"statusCode": 405, "headers": headers, "body": json.dumps({"error": "Method Not Allowed"})}

    try:
        body_raw = event.get("body", "{}")
        if isinstance(body_raw, (bytes, bytearray)):
            body_raw = body_raw.decode("utf-8")
        if event.get("isBase64Encoded") and body_raw:
            import base64 as _b64
            body_raw = _b64.b64decode(body_raw).decode("utf-8")
        payload = json.loads(body_raw or "{}")
    except Exception as e:
        headers["Content-Type"] = "application/json"
        return {"statusCode": 400, "headers": headers, "body": json.dumps({"error": f"Invalid JSON body: {e}"})}

    text = str(payload.get("text", "")).strip()
    voice = str(payload.get("voice", "")).strip()
    rate = str(payload.get("rate", "+0%")).strip() or "+0%"
    pitch = str(payload.get("pitch", "+0Hz")).strip() or "+0Hz"

    if not text:
        headers["Content-Type"] = "application/json"
        return {"statusCode": 400, "headers": headers, "body": json.dumps({"error": "Text is required"})}
    if not voice:
        headers["Content-Type"] = "application/json"
        return {"statusCode": 400, "headers": headers, "body": json.dumps({"error": "Voice is required"})}
    if len(text) > 2500:
        headers["Content-Type"] = "application/json"
        return {"statusCode": 413, "headers": headers, "body": json.dumps({"error": "Text quá dài (>2500 ký tự). Hãy chia thành các đoạn nhỏ hơn."})}

    try:
        audio_bytes = asyncio.run(_synthesize(text, voice, rate, pitch))
        if not audio_bytes or len(audio_bytes) < 100:
            headers["Content-Type"] = "application/json"
            return {"statusCode": 500, "headers": headers, "body": json.dumps({"error": "Không tạo được âm thanh (buffer quá nhỏ). Thử lại với giọng khác."})}
        headers["Content-Type"] = "audio/mpeg"
        headers["Cache-Control"] = "public, max-age=3600"
        return {
            "statusCode": 200,
            "headers": headers,
            "body": base64.b64encode(audio_bytes).decode("ascii"),
            "encoding": "base64",
        }
    except Exception as e:
        headers["Content-Type"] = "application/json"
        tb = traceback.format_exc()
        return {"statusCode": 500, "headers": headers, "body": json.dumps({"error": f"{e}", "trace": tb[:500]})}

File: api/tts.js
/**
 * Vercel Node.js Serverless Function: POST /api/tts
 * Spawns Python edge-tts via child_process (same pattern as local server.js)
 */
import { spawn } from 'child_process';

const PYTHON_SCRIPT = `
import sys, asyncio, json
import edge_tts
async def main():
    raw = sys.stdin.buffer.read().decode("utf-8")
    payload = json.loads(raw)
    t = payload["text"]
    v = payload["voice"]
    r = payload["rate"]
    p = payload["pitch"]
    comm = edge_tts.Communicate(t, v, rate=r, pitch=p)
    out_buf = bytearray()
    async for chunk in comm.stream():
        if chunk["type"] == "audio":
            out_buf.extend(chunk["data"])
    sys.stdout.buffer.write(bytes(out_buf))
    sys.stdout.buffer.flush()
asyncio.run(main())
`;

function getSafeParam(val, fallback, safePattern) {
  const s = String(val || fallback).trim();
  if (!s) return fallback;
  return s.replace(safePattern, '');
}

export default async function handler(req, res) {
  // CORS headers
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'GET,POST,OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type,Authorization');

  if (req.method === 'OPTIONS') {
    return res.status(200).end();
  }
  if (req.method !== 'POST') {
    return res.status(405).json({ error: 'Method Not Allowed. Use POST.' });
  }

  const { text, voice, rate, pitch } = req.body || {};
  const cleanText = String(text || '').trim();
  if (!cleanText) {
    return res.status(400).json({ error: 'Text is required' });
  }
  if (cleanText.length > 2500) {
    return res.status(413).json({ error: 'Text quá dài (>2500 ký tự). Hãy chia thành các đoạn nhỏ hơn.' });
  }

  const safeVoice = getSafeParam(voice, 'en-US-AriaNeural', /[^\w-]/g);
  const safeRate = getSafeParam(rate, '+0%', /[^\w%+\-.]/g);
  const safePitch = getSafeParam(pitch, '+0Hz', /[^\wHz+\-.]/g);

  try {
    const audioBuffer = await new Promise((resolve, reject) => {
      const child = spawn('python', ['-c', PYTHON_SCRIPT], {
        stdio: ['pipe', 'pipe', 'pipe'],
        maxBuffer: 100 * 1024 * 1024,
        env: { ...process.env, PYTHONIOENCODING: 'utf-8', PYTHONUTF8: '1' },
      });

      const chunks = [];
      const errOut = [];

      child.stdout.on('data', (d) => chunks.push(d));
      child.stderr.on('data', (d) => errOut.push(d));

      const timeout = setTimeout(() => {
        child.kill();
        reject(new Error('Timeout: Python TTS mất quá nhiều thời gian (>45s)'));
      }, 45000);

      child.on('close', (code) => {
        clearTimeout(timeout);
        if (code !== 0) {
          const errMsg = errOut.join('').slice(-3000);
          reject(new Error(`Python exited code ${code}: ${errMsg}`));
        } else if (chunks.length === 0) {
          reject(new Error('Không nhận được dữ liệu audio từ Python'));
        } else {
          resolve(Buffer.concat(chunks));
        }
      });

      child.on('error', (e) => {
        clearTimeout(timeout);
        reject(new Error(`Spawn error: ${e.message}`));
      });

      const payloadJson = JSON.stringify({
        text: cleanText,
        voice: safeVoice,
        rate: safeRate,
        pitch: safePitch,
      });
      child.stdin.write(Buffer.from(payloadJson, 'utf-8'));
      child.stdin.end();
    });

    if (audioBuffer.length < 100) {
      return res.status(500).json({ error: 'Không tạo được âm thanh (buffer quá nhỏ). Thử lại với giọng khác.' });
    }

    res.setHeader('Content-Type', 'audio/mpeg');
    res.setHeader('Cache-Control', 'public, max-age=3600');
    return res.status(200).send(audioBuffer);
  } catch (err) {
    return res.status(500).json({
      error: err.message,
      trace: err.stack ? err.stack.slice(0, 500) : '',
    });
  }
}